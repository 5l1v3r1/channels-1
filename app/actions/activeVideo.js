// @flow
import actionTypes from '../constants/actionTypes';
import type { ThunkAction, Dispatch, GetState } from '../constants/typeAliases';
import RatingType from '../dataTypes/ratingType';

export function updateActiveVideo(video: any): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    const { id } = getState().activeVideo;
    if (id !== video.id) {
      dispatch({
        type: actionTypes.UPDATE_ACTIVE_VIDEO,
        payload: video
      });
    }
  };
}

export function updateActiveVideoCounts(prevState: RatingType, rating: string): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    let likeCount = Number(getState().activeVideo.likeCount);
    let dislikeCount = Number(getState().activeVideo.dislikeCount);
    switch (rating) {
      case 'like':
        likeCount = prevState.like ? likeCount - 1 : likeCount + 1;
        dislikeCount = prevState.dislike ? dislikeCount - 1 : dislikeCount;
        break;
      case 'dislike':
        likeCount = prevState.like ? likeCount - 1 : likeCount;
        dislikeCount = prevState.dislike ? dislikeCount - 1 : dislikeCount + 1;
        break;
      default:
        likeCount = prevState.like ? likeCount - 1 : likeCount;
        dislikeCount = prevState.dislike ? dislikeCount - 1 : dislikeCount;
    }
    dispatch({
      type: actionTypes.UPDATE_ACTIVE_VIDEO,
      payload: {
        likeCount: String(likeCount),
        dislikeCount: String(dislikeCount)
      }
    });
  };
}
