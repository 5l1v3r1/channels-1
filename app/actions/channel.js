// @flow
import actionTypes from '../constants/actionTypes';
import type { Action } from '../constants/typeAliases';

export function selectChannel(id: string): Action {
  return {
    type: actionTypes.SELECT_CHANNEL,
    payload: id
  };
}
