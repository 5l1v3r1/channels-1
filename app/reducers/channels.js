// @flow
import { combineReducers } from 'redux';
import actionTypes from '../constants/actionTypes';
import type { Action, ChannelsState } from '../constants/typeAliases';
import ChannelType from '../dataTypes/channelType';
import { stringOrEmpty } from '../utils/utils';

function byId(state: { [string]: ChannelType } = {}, { type, payload }: Action): { [string]: ChannelType } {
  switch (type) {
    case actionTypes.FETCH_CHANNELS_SUCCESS:
      const nextState = { ...state };
      payload.forEach(item => {
        nextState[item.id] = ChannelType.from(item);
      });
      return nextState;
    default:
      return state;
  }
}

function allIds(state: string[] = [], { type, payload }: Action): string[] {
  switch (type) {
    case actionTypes.FETCH_CHANNELS_SUCCESS:
      return payload.map(item => stringOrEmpty(item.id));
    default:
      return state;
  }
}

function activeId(state: string = '', { type, payload }: Action): string {
  switch (type) {
    case actionTypes.UPDATE_ACTIVE_CHANNEL:
      return payload;
    default:
      return state;
  }
}

function isFetching(state: boolean = false, { type }: Action): boolean {
  switch (type) {
    case actionTypes.FETCH_CHANNELS_REQUEST:
      return true;
    case actionTypes.FETCH_CHANNELS_SUCCESS:
    case actionTypes.FETCH_CHANNELS_FAILURE:
      return false;
    default:
      return state;
  }
}

function error(state: boolean = false, { type }: Action): boolean {
  switch (type) {
    case actionTypes.FETCH_CHANNELS_REQUEST:
    case actionTypes.FETCH_CHANNELS_SUCCESS:
      return false;
    case actionTypes.FETCH_CHANNELS_FAILURE:
      return true;
    default:
      return state;
  }
}

const channels = combineReducers({
  byId,
  allIds,
  activeId,
  isFetching,
  error
});

export function getChannels(state: ChannelsState) {
  return state.allIds.map(id => state.byId[id]);
}

export function getActiveChannel(state: ChannelsState) {
  return state.byId[state.activeId];
}

export default channels;
