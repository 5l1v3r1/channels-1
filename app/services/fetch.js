// @flow
export const Methods = {
  GET: 'GET',
  POST: 'POST'
};

export function authRequest(uri: string): Promise<any> {
  const [url, body] = uri.split('?');
  const init = {
    method: Methods.POST,
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body
  };
  return fetch(url, init)
    .then(res => {
      if (!res.ok) {
        return Promise.reject();
      }
      return res.json();
    });
}

export function getRequest(uri: string): Promise<any> {
  const init = {
    method: Methods.GET,
    headers: { 'Content-Type': 'application/json' }
  };
  return fetch(uri, init)
    .then(res => {
      if (!res.ok) {
        return Promise.reject();
      }
      return res.json();
    });
}

export function postRequest(uri: string, body?: {}): Promise<any> {
  const init = body ?
  {
    method: Methods.POST,
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body)
  } :
  {
    method: Methods.POST,
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
  };
  return fetch(uri, init)
    .then(res => {
      if (!res.ok) {
        return Promise.reject();
      }
      return res;
    });
}
