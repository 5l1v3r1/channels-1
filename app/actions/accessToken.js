// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, Dispatch } from '../constants/typeAliases';
import { authRequest } from '../services/fetch';
import { getAccessTokenUri } from '../services/uriGenerator';

type ThunkAction = (dispatch: Dispatch) => any;

export function fetchAccessToken(): ThunkAction {
  return (dispatch: Dispatch) => {
    dispatch(requestAccessToken());
    const uri = getAccessTokenUri();
    return authRequest(uri)
      .then(json => dispatch(receiveAccessToken(json.access_token)))
      .catch(() => dispatch(accessTokenError()));
  };
}

export function requestAccessToken(): Action {
  return {
    type: actionTypes.FETCH_ACCESS_TOKEN_REQUEST
  };
}

export function receiveAccessToken(token: string): Action {
  return {
    type: actionTypes.FETCH_ACCESS_TOKEN_SUCCESS,
    payload: token
  };
}

export function accessTokenError(): Action {
  return {
    type: actionTypes.FETCH_ACCESS_TOKEN_FAILURE
  };
}
