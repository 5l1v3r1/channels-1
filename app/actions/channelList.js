// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, Dispatch, AccessTokenState } from '../constants/typeAliases';
import { apiRequest, Methods } from '../services/fetch';
import { getChannelsUri, getSubscriptionsUri } from '../services/uriGenerator';

type GetState = () => { accessToken: AccessTokenState };

type ThunkAction = (dispatch: Dispatch, getState: GetState) => any;

export function fetchChannels(): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    dispatch(requestChannels());
    const { accessToken } = getState().accessToken;
    const subscriptions = fetchSubscriptions(accessToken);
    subscriptions.then(json => {
      if (json.items == null) {
        return dispatch(receiveChannels([]));
      }
      const subscriptions = json.items.map(item => item.snippet.resourceId.channelId).join();
      const uri = getChannelsUri(accessToken, subscriptions);
      return apiRequest(uri, Methods.GET)
        .then(json => {
          if (json.items == null) {
            return dispatch(receiveChannels([]));
          }
          dispatch(receiveChannels(json));
        });
    })
    .catch(() => dispatch(channelsError()));
  };
}

function fetchSubscriptions(accessToken: string): Promise<any> {
  const uri = getSubscriptionsUri(accessToken);
  return apiRequest(uri, Methods.GET);
}

function requestChannels(): Action {
  return {
    type: actionTypes.REQUEST_CHANNELS
  };
}

function receiveChannels(json: any): Action {
  return {
    type: actionTypes.RECEIVE_CHANNELS,
    payload: json.items
  };
}

function channelsError(): Action {
  return {
    type: actionTypes.CHANNELS_ERROR
  };
}
