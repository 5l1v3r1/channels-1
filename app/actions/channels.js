// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, ThunkAction, Dispatch, GetState } from '../constants/typeAliases';
import { apiRequest, Methods } from '../services/fetch';
import { getChannelsUri, getSubscriptionsUri } from '../services/uriGenerator';

export function fetchChannels(): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    dispatch(requestChannels());
    const { token } = getState().accessToken;
    const subscriptions = fetchSubscriptions(token);
    return subscriptions.then(json => {
      if (json.items == null) {
        return dispatch(receiveChannels([]));
      }
      const subscriptions = json.items.map(item => item.snippet.resourceId.channelId).join();
      const uri = getChannelsUri(token, subscriptions);
      return apiRequest(uri, Methods.GET)
        .then(json => {
          if (json.items == null) {
            return dispatch(receiveChannels([]));
          }
          return dispatch(receiveChannels(json.items));
        });
    })
    .catch(() => dispatch(channelsError()));
  };
}

export function fetchSubscriptions(token: string): Promise<any> {
  const uri = getSubscriptionsUri(token);
  return apiRequest(uri, Methods.GET);
}

export function requestChannels(): Action {
  return {
    type: actionTypes.FETCH_CHANNELS_REQUEST
  };
}

export function receiveChannels(items: any): Action {
  return {
    type: actionTypes.FETCH_CHANNELS_SUCCESS,
    payload: items
  };
}

export function channelsError(): Action {
  return {
    type: actionTypes.FETCH_CHANNELS_FAILURE
  };
}

export function updateActiveChannel(channelId: string): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    const { activeId } = getState().channels;
    if (activeId !== channelId) {
      dispatch({
        type: actionTypes.UPDATE_ACTIVE_CHANNEL,
        payload: channelId
      });
    }
  };
}
