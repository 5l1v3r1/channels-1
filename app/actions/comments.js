// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, ThunkAction, Dispatch, GetState } from '../constants/types';
import { IComment } from '../dataTypes/commentType';
import { getRequest, postRequest } from '../services/fetch';
import { getTopLevelCommentsUri, postTopLevelCommentUri } from '../services/uriGenerator';

export function fetchComments(): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    dispatch(fetchCommentsRequest());
    const params = {
      videoId: getState().activeVideo.id,
      pageToken: getState().comments.pageToken
    };
    const uri = getTopLevelCommentsUri(params);
    return getRequest(uri)
      .then(
        json => {
          const { items, nextPageToken = params.pageToken } = json;
          return dispatch(fetchCommentsSuccess(items, nextPageToken));
        },
        () => dispatch(fetchCommentsFailure())
      );
  };
}

export function postComment(description: string): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    const body = {
      snippet: {
        channelId: getState().channels.activeId,
        videoId: getState().activeVideo.id,
        topLevelComment: {
          snippet: {
            textOriginal: description
          }
        }
      }
    };
    const uri = postTopLevelCommentUri();
    return postRequest(uri, body)
      .then(
        json => dispatch(postCommentSuccess(json)),
        () => Promise.reject()
      );
  };
}

export function fetchCommentsRequest(): Action {
  return {
    type: actionTypes.FETCH_COMMENTS_REQUEST
  };
}

export function fetchCommentsSuccess(items: IComment[], nextPageToken: string): Action {
  return {
    type: actionTypes.FETCH_COMMENTS_SUCCESS,
    payload: {
      items,
      nextPageToken
    }
  };
}

export function fetchCommentsFailure(): Action {
  return {
    type: actionTypes.FETCH_COMMENTS_FAILURE
  };
}

export function postCommentSuccess(json: IComment): Action {
  return {
    type: actionTypes.POST_COMMENT_SUCCESS,
    payload: json
  };
}
