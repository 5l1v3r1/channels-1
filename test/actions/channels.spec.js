/* eslint promise/always-return: 0 */
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import * as channelsActions from '../../app/actions/channels';
import * as fetch from '../../app/services/fetch';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock('../../app/services/uriGenerator');

describe('channels actions', () => {

  let store;
  beforeEach(() => {
    store = mockStore({
      accessToken: {
        token: 'ya29.GlyrBCJQJoIYFzocIunVN-CfjQZMG4oyVuAB6v_x_Z3FRnViyPy_deqRdwSAipQtKc9Nb2RudM9UISwI8SGNXxsJ1t3QHddeCdnoCjsM_vhLa9FlFVqMN_seI7oljg'
      },
      channels: {
        activeId: 'UCyIe-61Y8C4_o-zZCtO4ETQ'
      }
    });
  });

  it('creates FETCH_CHANNELS_SUCCESS when fetching subscriptions has resolved with undefined', () => {
    fetch.apiRequest = jest.fn(() => Promise.resolve({}));

    return store.dispatch(channelsActions.fetchChannels()).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it('creates FETCH_CHANNELS_SUCCESS when fetching subscriptions has resolved with null', () => {
    fetch.apiRequest = jest.fn(() => Promise.resolve({ items: null }));

    return store.dispatch(channelsActions.fetchChannels()).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it('creates FETCH_CHANNELS_FAILURE when fetching subscriptions has been rejected', () => {
    fetch.apiRequest = jest.fn(() => Promise.reject());

    return store.dispatch(channelsActions.fetchChannels()).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it('creates FETCH_CHANNELS_SUCCESS when fetching channels has been resolved', () => {
    fetch.apiRequest = jest.fn()
      .mockReturnValueOnce(Promise.resolve({
        items: [
          {
            snippet: {
              resourceId: {
                channelId: 'UCyIe-61Y8C4_o-zZCtO4ETQ'
              }
            }
          },
          {
            snippet: {
              resourceId: {
                channelId: 'UCO1cgjhGzsSYb1rsB4bFe4Q'
              }
            }
          }
        ]
      }))
      .mockReturnValueOnce(Promise.resolve({ items: ['UCyIe-61Y8C4_o-zZCtO4ETQ', 'UCO1cgjhGzsSYb1rsB4bFe4Q'] }));

    return store.dispatch(channelsActions.fetchChannels()).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it('creates FETCH_CHANNELS_FAILURE when fetching channels has been rejected', () => {
    fetch.apiRequest = jest.fn()
      .mockReturnValueOnce(Promise.resolve({
        items: [
          {
            snippet: {
              resourceId: {
                channelId: 'UCVTlvUkGslCV_h-nSAId8Sw'
              }
            }
          },
          {
            snippet: {
              resourceId: {
                channelId: 'UCW5YeuERMmlnqo4oq8vwUpg'
              }
            }
          }
        ]
      }))
      .mockReturnValueOnce(Promise.reject());

    return store.dispatch(channelsActions.fetchChannels()).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it('should create an action to request channels', () => {
    expect(channelsActions.requestChannels()).toMatchSnapshot();
  });

  it('should create an action to receive channels', () => {
    const items = ['UCyIe-61Y8C4_o-zZCtO4ETQ', 'UCO1cgjhGzsSYb1rsB4bFe4Q'];
    expect(channelsActions.receiveChannels(items)).toMatchSnapshot();
  });

  it('should create an action to handle an error', () => {
    expect(channelsActions.channelsError()).toMatchSnapshot();
  });

  it('should create an action to update the active channel', () => {
    const channelId = 'UCO1cgjhGzsSYb1rsB4bFe4Q';
    store.dispatch(channelsActions.updateActiveChannel(channelId));
    expect(store.getActions()).toMatchSnapshot();
  });
});
