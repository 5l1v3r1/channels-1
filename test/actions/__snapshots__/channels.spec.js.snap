// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`channels actions creates FETCH_CHANNELS_FAILURE when fetching channels has been rejected 1`] = `
Array [
  Object {
    "type": "FETCH_CHANNELS_REQUEST",
  },
  Object {
    "type": "FETCH_CHANNELS_FAILURE",
  },
]
`;

exports[`channels actions creates FETCH_CHANNELS_FAILURE when fetching subscriptions has been rejected 1`] = `
Array [
  Object {
    "type": "FETCH_CHANNELS_REQUEST",
  },
  Object {
    "type": "FETCH_CHANNELS_FAILURE",
  },
]
`;

exports[`channels actions creates FETCH_CHANNELS_SUCCESS when fetching channels has been resolved 1`] = `
Array [
  Object {
    "type": "FETCH_CHANNELS_REQUEST",
  },
  Object {
    "payload": Array [
      "UCyIe-61Y8C4_o-zZCtO4ETQ",
      "UCO1cgjhGzsSYb1rsB4bFe4Q",
    ],
    "type": "FETCH_CHANNELS_SUCCESS",
  },
]
`;

exports[`channels actions creates FETCH_CHANNELS_SUCCESS when fetching subscriptions has resolved with null 1`] = `
Array [
  Object {
    "type": "FETCH_CHANNELS_REQUEST",
  },
  Object {
    "payload": Array [],
    "type": "FETCH_CHANNELS_SUCCESS",
  },
]
`;

exports[`channels actions creates FETCH_CHANNELS_SUCCESS when fetching subscriptions has resolved with undefined 1`] = `
Array [
  Object {
    "type": "FETCH_CHANNELS_REQUEST",
  },
  Object {
    "payload": Array [],
    "type": "FETCH_CHANNELS_SUCCESS",
  },
]
`;

exports[`channels actions should create an action to handle an error 1`] = `
Object {
  "type": "FETCH_CHANNELS_FAILURE",
}
`;

exports[`channels actions should create an action to receive channels 1`] = `
Object {
  "payload": Array [
    "UCyIe-61Y8C4_o-zZCtO4ETQ",
    "UCO1cgjhGzsSYb1rsB4bFe4Q",
  ],
  "type": "FETCH_CHANNELS_SUCCESS",
}
`;

exports[`channels actions should create an action to request channels 1`] = `
Object {
  "type": "FETCH_CHANNELS_REQUEST",
}
`;

exports[`channels actions should create an action to update the active channel 1`] = `
Array [
  Object {
    "payload": "UCO1cgjhGzsSYb1rsB4bFe4Q",
    "type": "UPDATE_ACTIVE_CHANNEL",
  },
]
`;
